crates/walrus-core/src/encoding/blob_encoding.rs

```diff
@@ -24,38 +24,198 @@ use super::{
+        self.len() == 0
+    }
+}
+
+impl OwnedOrBorrowedBlob<'static> {
+    pub fn new_owned(blob: Vec<u8>) -> Self {
+        Self::Owned(blob)
+    }
+}
+
+impl<'a> AsRef<[u8]> for OwnedOrBorrowedBlob<'a> {
+    fn as_ref(&self) -> &[u8] {
+        match self {
+            Self::Borrowed(blob) => blob,
+            Self::Owned(blob) => blob,
+        }
+    }
+}
+
+/// Inner state of the blob encoder that doesn't contain the actual blob data.
+#[derive(Debug, Clone)]
+struct BlobEncoderData {
     /// The size of the encoded and decoded symbols.
     symbol_size: NonZeroU16,
     /// The number of rows of the message matrix.
     ///
     /// Stored as a `usize` for convenience, but guaranteed to be non-zero.
-    n_rows: usize,
+    n_rows: NonZeroU16,
     /// The number of columns of the message matrix.
     ///
     /// Stored as a `usize` for convenience, but guaranteed to be non-zero.
-    n_columns: usize,
+    n_columns: NonZeroU16,
     /// Reference to the encoding configuration of this encoder.
     config: EncodingConfigEnum,
     /// A tracing span associated with this blob encoder.
     span: Span,
 }
 
+impl BlobEncoderData {
+    fn get_encoder<E: EncodingAxis>(&self) -> ReedSolomonEncoder {
+        let EncodingConfigEnum::ReedSolomon(encoding_config) = self.config;
+        encoding_config
+            .get_encoder::<E>(self.sliver_length::<E::OrthogonalAxis>())
+            .expect("this length is compatible with the encoder")
+    }
+
+    /// Returns the size of the symbol in bytes.
+    pub fn symbol_usize(&self) -> usize {
+        self.symbol_size.get().into()
+    }
+
+    fn sliver_length<E: EncodingAxis>(&self) -> usize {
+        usize::from(self.config.n_source_symbols::<E::OrthogonalAxis>().get()) * self.symbol_usize()
+    }
+
+    fn empty_sliver<E: EncodingAxis>(&self, index: SliverIndex) -> SliverData<E> {
+        SliverData::<E>::new_empty(
+            self.config.n_source_symbols::<E::OrthogonalAxis>().get(),
+            self.symbol_size,
+            index,
+        )
+    }
+
+    fn empty_slivers<E: EncodingAxis>(&self) -> Vec<SliverData<E>> {
+        self.empty_slivers_range::<E>(0..self.config.n_shards().get())
+            .collect()
+    }
+
+    fn empty_slivers_range<E: EncodingAxis>(
+        &self,
+        range: Range<u16>,
+    ) -> impl Iterator<Item = SliverData<E>> {
+        range.map(|i| self.empty_sliver::<E>(SliverIndex(i)))
+    }
+
+    /// Returns a vector of empty [`SliverPair`] of length `n_shards`. Primary and secondary slivers
+    /// are initialized with the appropriate `symbol_size` and `length`.
+    #[tracing::instrument(level = "debug", skip_all)]
+    fn empty_sliver_pairs(&self) -> Vec<SliverPair> {
+        (0..self.config.n_shards().get())
+            .map(|i| SliverPair::new_empty(self.config, self.symbol_size, SliverPairIndex(i)))
+            .collect()
+    }
+
+    /// Computes the blob metadata from the provided leaf hashes.
```

Comment:
Nit:
```suggestion
    /// Computes the blob metadata from the provided leaf hashes of the symbols.
```

also mentioned that they are interpreted in the row (primary sliver) order?